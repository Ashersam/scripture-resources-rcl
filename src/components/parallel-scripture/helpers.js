import React, {forwardRef} from 'react';
import {
  AddBox,
  ArrowUpward,
  Check,
  ChevronLeft,
  ChevronRight,
  Clear,
  DeleteOutline,
  Edit,
  FilterList,
  FirstPage,
  LastPage,
  Remove,
  SaveAlt,
  Search,
  ViewColumn,
} from '@material-ui/icons';

export const tableIcons = {
  Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
  Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
  Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
  DetailPanel: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
  Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
  Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
  FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
  NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
  PreviousPage: forwardRef((props, ref) => <ChevronLeft {...props} ref={ref} />),
  ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
  SortArrow: forwardRef((props, ref) => <ArrowUpward {...props} ref={ref} />),
  ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
  ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />)
};

export const referenceIdsFromBooks = ({books}) => {
  const referenceIds = new Set([]);
  books.forEach((book) => {
    const bookIndex = {};
    Object.keys(book.chapters).forEach(chapterKey => {
      const chapter = book.chapters[chapterKey];
      Object.keys(chapter).forEach(verseKey => {
        const verse = chapter[verseKey];
        const referenceId = chapterKey + ':' + verseKey;
        bookIndex[referenceId] = verse;
        referenceIds.add(referenceId);
      });
    });
  });
  return [...referenceIds];
};

export const dataFromBooks = ({books}) => {
  const referenceIds = referenceIdsFromBooks({books});
  const data = referenceIds.map(referenceId => {
    let row = {referenceId};
    books.forEach((_, index) => {
      const [chapterKey, verseKey] = referenceId.split(':');
      const verse = books[index].chapters[chapterKey][verseKey];
      if (verse) row[index] = JSON.stringify({referenceId, ...verse});
    });
    return row;
  });
  return data;
};

export const dataFromReference = ({books, reference}) => {
  const referenceId = referenceIdFromReference(reference);
  let row = {referenceId};
  books.forEach((_, index) => {
    const verse = books[index].chapters[reference.chapter][reference.verse];
    if (verse) row[index] = JSON.stringify({referenceId, ...verse});
  });
  const data = [row];
  return data;
};

export const referenceIdFromReference = (reference) => reference.chapter + ':' + reference.verse;

export const referenceFromReferenceId = (referenceId) => {
  const [chapter, verse] = referenceId.split(':');
  return {chapter, verse};
};

