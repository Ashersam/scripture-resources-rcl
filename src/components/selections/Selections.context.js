import React, { useContext } from 'react'
import PropTypes, { func } from 'prop-types';

import { useSelections } from '..';

export const SelectionsContext = React.createContext({})

export function SelectionsContextProvider({
    selections,
    onSelections,
    occurrence,
    quote,
    onQuote,
    verseObjects,
    children,
  }) {
    const val = useSelections({     
        selections,
        onSelections,
        occurrence,
        quote,
        onQuote,
        verseObjects,
    });
  
    return (
      <SelectionsContext.Provider value={val}>
        {children}
      </SelectionsContext.Provider>
    );
  };
  
  SelectionsContextProvider.propTypes = {
    /** words in a selection */
    selections: PropTypes.array,
    /** action taken after a selection is made */
    onSelections: PropTypes.func.isRequired,
    /** the quote to be selected */
    quote: PropTypes.string.isRequired,
    /** the verses where quote may be found */
    verseObjects: PropTypes.array,
    /** if quote occurs mulitple times, this is the occurence of the one selected */
    occurrence: PropTypes.number,
    /** action taken when quote is provided */
    onQuote: PropTypes.func,
  };
  
  export default SelectionsContextProvider;
