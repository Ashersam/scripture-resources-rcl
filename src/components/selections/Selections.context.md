```js
import ReactJson from 'react-json-view';
import { SelectionsContext, SelectionsContextProvider } from "scripture-resources-rcl";

function Component () {
  const selectionsContext = React.useContext(SelectionsContext);

  return (
    <div>
      <ReactJson src={selectionsContext} />
    </div>
  );
};


const quote = "ὁ λόγος,";
//const onSelections = () => console.log("hello");
const occurrence = 1;
const verseObjects = [
  {
    "text": "ἐν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἀρχῇ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἦν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ὁ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "λόγος",
    "type": "word",
  },
  {
    "type": "text",
    "text": ", "
  },
  {
    "text": "καὶ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ὁ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "λόγος",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἦν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "πρὸς",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "τὸν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "Θεόν",
    "type": "word",
  },
  {
    "type": "text",
    "text": ", "
  },
  {
    "text": "καὶ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "Θεὸς",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἦν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ὁ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "λόγος",
    "type": "word",
  },
  {
    "type": "text",
    "text": ".\n\n"
  }
];

const props = {
  quote,
  occurrence,
  verseObjects,
};


<div style={{height: '250px', overflow: 'auto'}}>
  <SelectionsContextProvider value={{...props}}>
    <Component />
  </SelectionsContextProvider>
</div>
```
