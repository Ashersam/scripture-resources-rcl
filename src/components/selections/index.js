export { default as useSelections } from './useSelections';
export * from './Selections.context';