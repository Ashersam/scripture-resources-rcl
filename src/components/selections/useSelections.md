withSelections example:

```js
import ReactJson from 'react-json-view';
import { useSelections } from 'scripture-resources-rcl';

function Component ({
  quote,
  occurrence,
  verseObjects,
}) {
  const [selections, setSelections] = React.useState();
  const { state, actions } = useSelections({
    selections,
    onSelections: setSelections,
    quote,
    occurrence,
    verseObjects,
  });

  return (
    <ReactJson src={state} />
  );
};

const quote = "ὁ λόγος,";
const occurrence = 1;
const verseObjects = [
  {
    "text": "ἐν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἀρχῇ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἦν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ὁ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "λόγος",
    "type": "word",
  },
  {
    "type": "text",
    "text": ", "
  },
  {
    "text": "καὶ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ὁ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "λόγος",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἦν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "πρὸς",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "τὸν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "Θεόν",
    "type": "word",
  },
  {
    "type": "text",
    "text": ", "
  },
  {
    "text": "καὶ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "Θεὸς",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ἦν",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "ὁ",
    "type": "word",
  },
  {
    "type": "text",
    "text": " "
  },
  {
    "text": "λόγος",
    "type": "word",
  },
  {
    "type": "text",
    "text": ".\n\n"
  }
];

const props = {
  quote,
  occurrence,
  verseObjects,
};

<Component {...props} />;
```