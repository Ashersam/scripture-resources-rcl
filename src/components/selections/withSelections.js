import React, { useState, useEffect } from "react";
import PropTypes from 'prop-types';
import deepFreeze from 'deep-freeze';

// const stringify = (array) => array.map(object => JSON.stringify(object));
const parsify = (array) => array.map(string => JSON.parse(string));

function withSelections(Component){
  return function SelectionsComponent({
    quote,
    onQuote,
    quoteVerseObjects,
    ...props
  }) {
    const [selections, setSelections] = useState([]);

    // const wordsFromMilestone = ({milestone, quotedWords=[]}) => {
    //   const _selections = selections.map(selection => selection.text);
    //   if (_selections.include(milestone.content)) {
    //     if (milestone.children[0].type === 'word') {
    //       const words = milestone.children;
    //       words.forEach(word => quotedWords.push(word.text));
    //     }
    //   }
    // };

    useEffect(() => {
      const quoteFromVerse = () => {
        let quotedWords = new Array(quoteVerseObjects.length);
        const _selections = selections.map(selection => JSON.parse(selection).text);
        quoteVerseObjects.forEach((verseObject, index) => {
          const {type, text} = verseObject;
          if (type === 'word') {
            const match = _selections.includes(text);
            const quotedWord = match ? text : '…';
            quotedWords.push(quotedWord);
          }
        });
        const quote = quotedWords.join(' ')
        .replace(/( ?… ?)+/g,' … ').replace(/(^[… ]+|[… ]+$)/g, '');
        return quote;
      };

      if (quoteVerseObjects && onQuote) {
        const quote = quoteFromVerse();
        onQuote(quote);
      }
    }, [selections, onQuote, quoteVerseObjects]);

    const selectionFromWord = (word) => {
      const {content, text} = word;
      const selection = JSON.stringify({text: content || text});
      return selection;
    };

    const isSelected = (word) => {
      const selection = word => selectionFromWord(word);
      const selected = selections.includes(selection);
      return selected;
    };

    const areSelected = (words) => {
      let selected = false;
      const _selections = words.map(word => selectionFromWord(word));
      _selections.forEach(selection => {
        if (selections.includes(selection)) selected = true;
      });
      return selected;
    };

    const addSelection = (word) => {
      let _selections = new Set(selections);
      const selection = selectionFromWord(word);
      _selections.add(selection);
      setSelections([..._selections]);
    };

    const addSelections = (words) => {
      let _selections = new Set(selections);
      words.forEach(word => {
        const selection = selectionFromWord(word);
        _selections.add(selection);
      });
      setSelections([..._selections]);
    };

    const removeSelection = (word) => {
      const selection = selectionFromWord(word);
      const _selections = new Set(selections);
      _selections.delete(selection);
      setSelections([..._selections]);
    };

    const removeSelections = (words) => {
      let _selections = new Set(selections);
      words.forEach(word => {
        const selection = selectionFromWord(word);
        _selections.delete(selection);
      });
      setSelections([..._selections]);
    };

    const _selections = deepFreeze(parsify(selections));

    const value = {
      selections: _selections,
      isSelected,
      areSelected,
      addSelection,
      addSelections,
      removeSelection,
      removeSelections,
    };

    return (
      <Component value={value} {...props} />
    );
  }
}

withSelections.propTypes = {
  quoteVerseObjects: PropTypes.array,
  onQuote: PropTypes.func,
};

withSelections.defaultProps = {
  selections: [],
};

export default withSelections;
