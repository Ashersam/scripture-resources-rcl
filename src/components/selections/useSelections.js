import React, { useCallback, useEffect } from "react";
import PropTypes from 'prop-types';
import deepFreeze from 'deep-freeze';

import helpers, {parsify, selectionsFromQuote, quoteFromVerse} from './helpers';

function useSelections({
  selections,
  onSelections,
  occurrence,
  quote,
  onQuote,
  verseObjects,
}) {

  useEffect(() => {
    const _selections = selectionsFromQuote({quote, verseObjects, occurrence});
    update(_selections);
  }, [quote, occurrence, verseObjects]);

  useEffect(() => {
    if (verseObjects && onQuote) {
      const _quote = quoteFromVerse({selections, verseObjects});
      onQuote(_quote);
    }
  }, [selections, onQuote, verseObjects]);

  const update = useCallback((_selections) => {
    const __selections = _selections && deepFreeze(parsify(_selections));
    onSelections(__selections);
  }, [onSelections]);

  const isSelected = (word) => helpers.isSelected({word, selections});

  const areSelected = (words) => helpers.areSelected({words, selections});

  const addSelection = (word) => {
    let _selections = helpers.addSelection({word, selections});
    update(_selections);
  };

  const addSelections = (words) => {
    let _selections = helpers.addSelections({words, selections});
    update(_selections);
  };

  const removeSelection = (word) => {
    const _selections = helpers.removeSelection({word, selections});
    update(_selections);
  };

  const removeSelections = (words) => {
    let _selections = helpers.removeSelections({words, selections});
    update(_selections);
  };

  return {
    state: selections,
    actions: {
      update,
      isSelected,
      areSelected,
      addSelection,
      addSelections,
      removeSelection,
      removeSelections,
    },
  };
};

useSelections.propTypes = {
  selections: PropTypes.array,
  onSelections: PropTypes.func.isRequired,
  quote: PropTypes.string.isRequired,
  verseObjects: PropTypes.array,
  occurrence: PropTypes.number,
  onQuote: PropTypes.func,
};

export default useSelections;
