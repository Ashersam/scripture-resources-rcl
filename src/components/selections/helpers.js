import { selectionsFromQuoteAndVerseObjects } from '../../core/selections/selections';

// const stringify = (array) => array.map(object => JSON.stringify(object));
//export const parsify = (array) => array.map(string => JSON.parse(string));

export const selectionsFromQuote = ({ quote, verseObjects, occurrence }) => {
  let selections = [];
  if (quote && verseObjects && occurrence) {
    selections = selectionsFromQuoteAndVerseObjects({ quote, verseObjects, occurrence })
      .map(selection => JSON.stringify(selection));
  }
  return selections;
};

export const quoteFromVerse = ({ selections, verseObjects }) => {
  let quotedWords = new Array(verseObjects.length);
  const _selections = selections.map(selection => JSON.parse(selection).text);
  verseObjects.forEach((verseObject, index) => {
    const { type, text } = verseObject;
    if (type === 'word') {
      const match = _selections.includes(text);
      const quotedWord = match ? text : '…';
      quotedWords.push(quotedWord);
    }
  });
  const quote = quotedWords.join(' ')
    .replace(/( ?… ?)+/g, ' … ').replace(/(^[… ]+|[… ]+$)/g, '');
  return quote;
};

// const wordsFromMilestone = ({milestone, quotedWords=[]}) => {
//   const _selections = selections.map(selection => selection.text);
//   if (_selections.include(milestone.content)) {
//     if (milestone.children[0].type === 'word') {
//       const words = milestone.children;
//       words.forEach(word => quotedWords.push(word.text));
//     }
//   }
// };

export const selectionFromWord = (word) => {
  const { content, text, occurrence, occurrences } = word;
  const selectionObject = {
    text: content || text,
    occurrence: parseInt(occurrence),
    occurrences: parseInt(occurrences),
  };
  return selectionObject;
};

export const isSelected = ({ word, selections }) => {
  const selection = selectionFromWord(word);
  debugger;
  const selected = selections.includes(selection);
  return selected;
};

export const areSelected = ({ words, selections }) => {
  let selected = false;
  const _selections = words.slice(0);
  debugger;
  _selections.forEach(selection => {
    //if (selections.includes(_s)) selected = true;
    const _selection = JSON.parse(selection);
    let _text = _selection.text;
    let _occ  = _selection.occurrence;
    let _occs = _selection.occurrences;
    for (let i=0; i < selections.length; i++) {
      const text = selections[i].text;
      const occ  = selections[i].occurrence;
      const occs = selections[i].occurrences;
      if ( text === _text &&  occ === _occ && occs === _occs) {
        selected = true;
        break;
      }
    }
  });
  return selected;
};

export const addSelection = ({ word, selections }) => {
  let _selections = selections.slice(0);
  const selection = selectionFromWord(word);
  debugger;
  _selections.add(selection);
  return [..._selections];
};

export const addSelections = ({ words, selections }) => {
  let _selections = selections.slice(0);
  words.forEach(word => {
    const selection = selectionFromWord(word);
    debugger;
    _selections.add(selection);
  });
  return [..._selections];
};

export const removeSelection = ({ word, selections }) => {
  const selection = selectionFromWord(word);
  debugger;
  const _selections = selections.slice(0);
  _selections.delete(selection);
  return [..._selections];
};

export const removeSelections = ({ words, selections }) => {
  debugger;
  let _selections = selections.slice(0);
  words.forEach(word => {
    const selection = selectionFromWord(word);
    debugger;
    _selections.delete(selection);
  });
  debugger;
  return [..._selections];
};

export default {
  selectionFromWord,
  isSelected,
  areSelected,
  addSelection,
  addSelections,
  removeSelection,
  removeSelections,
};