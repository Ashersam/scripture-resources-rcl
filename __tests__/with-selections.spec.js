import { selectionsFromQuote } from '../src/components/selections/helpers';
import usfmJS from 'usfm-js';
import fs from 'fs-extra';
import hi_tit from '../src/components/mocks/ugnt_tit.usfm';

describe('Selections From Quote', () => {
  const output = [];
  it('should output', () => {
    const bookName = 'titus';
    const book = usfmJS.toJSON(hi_tit);
    const { chapters } = book;
    const chaptersNumber = Object.keys(chapters).length - 1;
    for (let i = 1; i <= chaptersNumber; i++) {
      const versesNumber = Object.keys(chapters[i]).length - 1;
      for (let j = 1; j <= versesNumber; j++) {
        const { verseObjects } = chapters[i][j];
        const quote = verseObjects.map(({ text }) => text).join('');
        const occurrence = 1;
        const selections = selectionsFromQuote({ quote, verseObjects, occurrence });
        const selectionsString = selections.map((string) => {
          const selection = JSON.parse(string);
          return selection.text;
        }).join(' ');
        if (selectionsString !== quote) {
          output.push({
            book: bookName,
            chapter: i,
            verse: j,
            quote,
            highlight: selectionsString
          })
        }
      }
    }
    fs.writeJSONSync(`./${bookName}-expected-highlights.json`, output, { spaces: '\t' });
  })
})