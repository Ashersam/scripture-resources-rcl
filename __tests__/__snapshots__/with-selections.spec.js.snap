// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selections From Quote should output 1`] = `
Array [
  "{\\"text\\":\\"τῆς\\",\\"occurrence\\":1,\\"occurrences\\":1}",
  "{\\"text\\":\\"κατ\\",\\"occurrence\\":2,\\"occurrences\\":2}",
  "{\\"text\\":\\"εὐσέβειαν\\",\\"occurrence\\":1,\\"occurrences\\":1}",
]
`;
